### ISSUE TITLE
Semantic facet optimization idea: estimate for all possible facet values

### ISSUE THREAD

**Issue by brent-hartwig** [project-lux/lux-marklogic#365]
**Problem Description:**
Copied from #362 

> Semantic facet requests can timeout, even after the 59 seconds they are allotted. Unknown how likely users would use these facets should they take more than n seconds to appear.
>
> An extreme example takes no less than 4.4 minutes: the responsible collections facet for an estimated ~15 million items curated by Yale University Library (YUL) --a boil the ocean search that LUX is asked to calculate facets for. Using the techniques described below, this facet request completed in 10 seconds.

**Expected Behavior/Solution:**
For the two semantic facets we have today (responsible units and collections):

1. Start with all possible facet values.  Collections is the largest with 64.
2. Incorporating the user's search criteria, calculate the estimate for each of the facet values.
3. Filter out facet values that have an estimate of zero values.

Results for the extreme example using the above technique:

`11.3.0-q17-cts-speed-demon firstRun=2815 warmRuns=10 warmMin=238 warmMax=973 warmAvg=332 stddev=215 totalItemsRead=198`

All durations are in milliseconds, meaning the slowest it becomes to calculate the responsible collections facet for ~15 million items curated by YUL is 2.815 seconds 💥 

Results for the responsible collections facet for 675K items matching the "journals" keyword search:

`11.3.0-q17-cts-speed-demon firstRun=1684 warmRuns=10 warmMin=149 warmMax=166 warmAvg=154 stddev=5 totalItemsRead=572`

Yep, 1.7 seconds with empty caches :)

**Once implemented, I don't believe a semantic facet request would ever time out.**

Implementation, within the context of the CTS benchmark template:

[q17-cts-speed-demon.js.txt](https://github.com/user-attachments/files/17573881/q17-cts-speed-demon.js.txt)

CTS queries for both examples are included.  Set `baseSearchCriteria` to `journalsBaseSearchCriteria` or `curatedByYulBaseSearchCriteria`.

To implement within LUX, we'd need to update the search criteria within facetsViaSearchConfig.mjs and the approach within facetsLib.mjs.

**Requirements:**
See above.

**Needed for promotion:**
If an item on the list is not needed, it should be crossed off but not removed.

~- [ ] Wireframe/Mockup - Mike~
- [ ] Committee discussions - Sarah
- [ ] Feasibility/Team discussion - Sarah
- [ ] Backend requirements - See the "Expected Behavior/Solution" section.
~- [ ] Frontend requirements - TBD~
- [ ] Are new regression tests required for QA - Amy
- [ ] Questions
- List of questions for discussions. Answers should be documented within the issue.

**UAT/LUX Examples:**

- [~15m items curated by YUL](https://lux-front-tst.collections.yale.edu/view/results/objects?q=%7B%22OR%22%3A%5B%7B%22memberOf%22%3A%7B%22curatedBy%22%3A%7B%22id%22%3A%22https%3A%2F%2Flux.collections.yale.edu%2Fdata%2Fgroup%2Fd07b9b86-0a1e-4026-aa4c-8ecba8bbd9c9%22%7D%7D%7D%2C%7B%22memberOf%22%3A%7B%22curatedBy%22%3A%7B%22memberOf%22%3A%7B%22id%22%3A%22https%3A%2F%2Flux.collections.yale.edu%2Fdata%2Fgroup%2Fd07b9b86-0a1e-4026-aa4c-8ecba8bbd9c9%22%7D%7D%7D%7D%5D%7D)
- [~675K items matching "journals" keyword search](https://lux-front-tst.collections.yale.edu/view/results/objects?q=%7B%22text%22%3A%22journals%22%7D)

**Dependencies/Blocks:**

This ticket neither blocks nor is blocked by another ticket.

**Related Github Issues:**

- #362: a slower optimization of the semantic facets, and would impose implementation complexities.

**Related links:**

- These links can consist of resources, bugherds, etc.

**Wireframe/Mockup:**

N/A

**Comment by brent-hartwig** [comment]
Updated q17-cts-speed-demon.js.txt after having compared the performance of the universal index and a temporary field for the collection set criteria.  In this case, the universal index proved faster.  The updated version of the query allows you to try either, so long as the field exists.  The field's path is /json[type = "Set" and classified_as/equivalent/id = "http://vocab.getty.edu/aat/300025976"]/id.

Comparison:

11.3.0-q17-cts-universal-index firstRun=2422 warmRuns=10 warmMin=237 warmMax=289 warmAvg=256 stddev=18 totalItemsRead=198

11.3.0-q17-cts-field-range-index firstRun=2510 warmRuns=10 warmMin=301 warmMax=716 warmAvg=352 stddev=122 totalItemsRead=198

**Comment by clarkepeterf** [comment]
@brent-hartwig @azaroth42 

This method works well for Collections since its easy to get a list of all collection sets by their classification (http://vocab.getty.edu/aat/300025976). In order to extend this approach to Responsible Units, we need a way of getting a list of all Units.

There are two ways to do this:

1 )  Try to derive which Groups are Units based on a query. This allows would auto-update the list of units based on the dataset. However, I can't find any classifications or JSON properties to make this possible.
  - For the original 4 units (YPM, YUAG, YCBA, YUL), we can search for "has a member Group which is classified as a Department (http://vocab.getty.edu/aat/300263534)"
    - But this still excludes PMC, LML, Elizabethan Club, so we need an alternate method for finding those

In order to use method 1, I need a method to determine a list of all units with a query. However, there are no obvious classifications or JSON properties to use to get this list.


2 ) Include IRI constants for each Unit in our code. This will require updates to our code if any additional Units are added to our data.

This is a good solution if:
 a. It is unlikely we will be adding new Units, or it will be rare.
 b. The IRIs listed below are expected to remain constant across datasets
   - http://vocab.getty.edu/ulan/500303561 (YPM)
   - http://vocab.getty.edu/ulan/500303559 (YUAG)
   - http://vocab.getty.edu/ulan/500303557 (YCBA)
   - http://vocab.getty.edu/ulan/500303563 (YUL)
   - https://data.paul-mellon-centre.ac.uk/group/pmc (PMC)
   - https://paperbase.xyz/records/lml.json (LML)
   - https://linked-art.library.yale.edu/node/e53eeb74-41a9-4eb3-bc0d-f908fa6da183 (Elizabethan Club)


So, in summary:

Am I missing anything that would make option 1 possible?
Do we think the IRIs are constant and adding units is unlikely enough to go with option 2?

**Comment by brent-hartwig** [comment]
@clarkepeterf, would it take long to query for the responsible units of the 64 collections?  Alternatively, perhaps @azaroth42 could add a responsible unit equivalent of collection's http://vocab.getty.edu/aat/300025976 classification, in support of your first idea.  My gut says the second idea is a workaround that could bite us later.

**Comment by clarkepeterf** [comment]
@brent-hartwig Yes - that works! Thanks!

**Comment by roamye** [comment]
The UAT links seem to be faster in TST. (used my clock timer so it was hard to start/stop at the correct times) 
[items matching "books"](https://lux-front-tst.collections.yale.edu/view/results/objects?q=%7B%22text%22%3A%22books%22%2C%22_lang%22%3A%22en%22%7D&sq=books)
[~15m items curated by YUL](https://lux-front-tst.collections.yale.edu/view/results/objects?q=%7B%22OR%22%3A%5B%7B%22memberOf%22%3A%7B%22curatedBy%22%3A%7B%22id%22%3A%22https%3A%2F%2Flux.collections.yale.edu%2Fdata%2Fgroup%2Fd07b9b86-0a1e-4026-aa4c-8ecba8bbd9c9%22%7D%7D%7D%2C%7B%22memberOf%22%3A%7B%22curatedBy%22%3A%7B%22memberOf%22%3A%7B%22id%22%3A%22https%3A%2F%2Flux.collections.yale.edu%2Fdata%2Fgroup%2Fd07b9b86-0a1e-4026-aa4c-8ecba8bbd9c9%22%7D%7D%7D%7D%5D%7D)
[~675K items matching "journals" keyword search](https://lux-front-tst.collections.yale.edu/view/results/objects?q=%7B%22text%22%3A%22journals%22%7D)

**Comment by brent-hartwig** [comment]
@roamye, you can use the browser's developer tools to see the response times of individual requests.  I also entered "responsible" in the filter box so as to cut down on the noise:

![Image](https://github.com/user-attachments/assets/ef563106-ce04-409d-8ba4-1100b953a2c8)

Additional performance comparisons were captured in the [Semantic Facet Optimization in Data Dev](https://teams.microsoft.com/l/message/19:351e0281aaa144d0812b960d216958c6@thread.tacv2/1737655021745?tenantId=dd8cbebb-2139-4df8-b411-4e3e87abeb5c&groupId=4620eac6-0955-4c69-93e8-b493b21f0df6&parentMessageId=1737655021745&teamName=ML%2F%20Cultural%20Heritage%20Support&channelName=1%20-%20LUX_ML&createdTime=1737655021745) thread.

Be sure to use change up your search criteria when testing in Blue and Green!

**Comment by roamye** [comment]
Approved by UAT
